src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java

```diff
@@ -93,10 +105,46 @@ private JsonArray formatBatch(String topic) {
 
             JsonElement recordJSON = recordToJSON(record);
             JsonObject message = populateMetadata(topic, recordJSON);
-            batchRecords.add(message);
+            JsonArray currentBatchRecords = batchRecordsList.get(currentBatchListIndex);
+
+            try {
+                if (hasBatchOverflowed(message, currentBatchRecords)) {
+                    JsonArray emptyBatch = new JsonArray();
+
+                    emptyBatch.add(message);
+                    batchRecordsList.add(currentBatchListIndex++, emptyBatch);
+                } else {
+                    currentBatchRecords.add(message);
+                }
+
+            } catch (SizeLimitExceededException ex) {
+                log.error("Single message exceeds size limit.", ex);
+            }
         }
+        return batchRecordsList;
+    }
 
-        return batchRecords;
+    private boolean hasBatchOverflowed(JsonObject jsonRecord, JsonArray
```

Comment:
I'm concerned that this implementation is fairly inefficient in that it serializes the batch to determine if it is oversized, but then throws that serialization away. I think serialization is indeed required to determine if the batch is oversized, but I'd like to see the implementation use the serialized batch to make the common case, where it doesn't need to be split, more efficient.